README.md
# Catenis Off-Chain Library

A Node.js library used for manipulating Catenis off-chain message data structures.

## License

This software is released under the [MIT License](LICENSE). Feel free to fork, and modify!

Copyright © 2020-2023, Blockchain of Things Inc.


CHANGELOG.md
# Changelog

## [1.2.6] - 2023-03-03

### Changes
- Update development dependency modules.

### Fix
- Make code compatible with the browser implementation of Buffer, so that the library may also be used in web browsers.
- Update dependency modules to mitigate security vulnerabilities

## [1.2.5] - 2023-02-09

### Changes
- No source code changes; update package.json so dependency modules not published on NPM can be freely downloaded.

## [1.2.4] - 2023-01-26

### Changes
- No source code changes; updates only to license, copyright and source code repository information.

## [1.2.3] - 2021-05-26

### Changes
- Update dependency modules to be in sync with the latest versions of IPFS components.

## [1.2.2] - 2021-05-13

### Fixes
- Update dependency modules to mitigate security vulnerabilities.

## [1.2.1] - 2020-08-01

### Fixes
- Update dependency modules to mitigate security vulnerabilities.

## [1.2.0] - 2020-06-12

### New features
- Support for a new version (ver. 1) of the Batch Document, which has a different implementation of the Merkle tree.

## [1.1.0] - 2020-05-26

### Changes
- Drop use of features that are only available in newer versions of the Buffer class so it is compatible with the
 browser implementation of that class thus allowing for this library to be used from within a browser.

## [1.0.0] - 2020-01-16

### New features
- Initial version.


test/testUtil.js
const expect = require('chai').expect;
const Util = require('../src/Util');

describe('Write 64-bit numbers to buffer', function () {
    const buf64 = Buffer.alloc(8);

    it('should correctly write small numbers (< 32 bits)', function () {
        Util.writeInt64BE(12345, buf64, 0);

        expect(buf64.equals(Buffer.from('0000000000003039', 'hex'))).to.be.true;
    });

    it('should correctly write large numbers (> 32 bits)', function () {
        Util.writeInt64BE(6336313684422, buf64, 0);

        expect(buf64.equals(Buffer.from('000005c349b9f1c6', 'hex'))).to.be.true;
    });

    it('should correctly write small negative numbers (< 32 bits)', function () {
        Util.writeInt64BE(-12345, buf64, 0);

        expect(buf64.equals(Buffer.from('ffffffffffffcfc7', 'hex'))).to.be.true;
    });

    it('should correctly write large negative numbers (> 32 bits)', function () {
        Util.writeInt64BE(-6336313684422, buf64, 0);

        expect(buf64.equals(Buffer.from('fffffa3cb6460e3a', 'hex'))).to.be.true;
    });

    it('should correctly write negative numbers multiple of 2 to the power of 32 (> 32 bits)', function () {
        Util.writeInt64BE(-1408070668255232, buf64, 0);

        expect(buf64.equals(Buffer.from('fffaff5e00000000', 'hex'))).to.be.true;
    });
});

describe('Read 64-bit numbers from buffer', function () {
    const buf64 = Buffer.alloc(8);

    it('should correctly read small numbers (< 32 bits)', function () {
        Buffer.from('0000000000003039', 'hex').copy(buf64);

        expect(Util.readInt64BE(buf64, 0)).to.equal(12345);
    });

    it('should correctly read large numbers (> 32 bits)', function () {
        Buffer.from('000005c349b9f1c6', 'hex').copy(buf64);

        expect(Util.readInt64BE(buf64, 0)).to.equal(6336313684422);
    });

    it('should correctly read small negative numbers (< 32 bits)', function () {
        Buffer.from('ffffffffffffcfc7', 'hex').copy(buf64);

        expect(Util.readInt64BE(buf64, 0)).to.equal(-12345);
    });

    it('should correctly read large negative numbers (> 32 bits)', function () {
        Buffer.from('fffffa3cb6460e3a', 'hex').copy(buf64);

        expect(Util.readInt64BE(buf64, 0)).to.equal(-6336313684422);
    });

    it('should correctly write negative numbers multiple of 2 to the power of 32 (> 32 bits)', function () {
        Buffer.from('fffaff5e00000000', 'hex').copy(buf64);

        expect(Util.readInt64BE(buf64, 0)).to.equal(-1408070668255232);
    });
});


src/Util.js
const CID = require('cids');
const BigNumber = require('bignumber.js');
const bitcoinLib = require('bitcoinjs-lib');

const MAX_TIMESTAMP = 8640000000000000;     // 100,000,000 years (in milliseconds) after Unix's epoch
const MIN_TIMESTAMP = -MAX_TIMESTAMP;       // 100,000,000 years (in milliseconds) before Unix's epoch
const COMPRESSED_PUB_KEY_LENGTH = 33;

module.exports = {
    dummyPubKeyHash: Buffer.alloc(20, 0x00),
    validatePubKeyHash(pubKeyHash) {
        let bufPubKeyHash;

        if (Buffer.isBuffer(pubKeyHash)) {
            bufPubKeyHash = pubKeyHash;
        }
        else if (typeof pubKeyHash === 'string') {
            // Public key hash should be base64 encoded. Try to decode it
            const decPubKeyHash = Buffer.from(pubKeyHash, 'base64');

            if (decPubKeyHash.toString('base64') === pubKeyHash) {
                bufPubKeyHash = decPubKeyHash;
            }
        }

        if (bufPubKeyHash && bufPubKeyHash.byteLength === 20) {
            return bufPubKeyHash;
        }
    },
    isValidTimestamp(timestamp) {
        return Number.isInteger(timestamp) && timestamp >= MIN_TIMESTAMP && timestamp <= MAX_TIMESTAMP;
    },
    validateCid(cid) {
        if (CID.isCID(cid)) {
            return cid;
        }

        let validCid;

        try {
            validCid = new CID(cid);
        }
        catch (err) {}

        return validCid;
    },
    keyPairFromPublicKey(pubKey) {
        const opts = pubKey.byteLength > COMPRESSED_PUB_KEY_LENGTH ? {compressed: false} : undefined;

        return bitcoinLib.ECPair.fromPublicKey(pubKey, opts);
    },
    writeUInt64BE(num, buf, offset) {
        const bnNum = new BigNumber(num);
        const bnUpperValue = bnNum.dividedToIntegerBy(0x100000000);
        offset = buf.writeUInt32BE(bnUpperValue.toNumber(), offset);
        offset = buf.writeUInt32BE(bnNum.minus(bnUpperValue.times(0x100000000)).toNumber(), offset);

        return offset;
    },
    readUInt64BE(buf, offset) {
        const bnUpperValue = new BigNumber(buf.readUInt32BE(offset));
        return bnUpperValue.times(0x100000000).plus(buf.readUInt32BE(offset + 4)).toNumber();
    }
};

module.exports.writeInt64BE = function writeInt64BE(num, buf, offset) {
    if (num >= 0) {
        return module.exports.writeUInt64BE(num, buf,offset);
    }

    num = -num;

    const bnNum = new BigNumber(num);
    const bnUpperValue = bnNum.dividedToIntegerBy(0x100000000);
    const bnLowerValue = bnNum.minus(bnUpperValue.times(0x100000000));

    if (bnLowerValue.isZero()) {
        // Write 2's complement (i.e. -n) of upper value
        offset = buf.writeInt32BE(bnUpperValue.negated().toNumber(), offset);
    }
    else {
        // Write 1's complement of upper value
        offset = buf.writeUInt32BE(new BigNumber('ffffffff', 16).minus(bnUpperValue).toNumber(), offset);
    }

    // Write 2's complement (i.e. -n) of lower value
    offset = buf.writeInt32BE(bnLowerValue.negated().toNumber(), offset);

    return offset;
};

module.exports.readInt64BE = function readInt64BE(buf, offset) {
    if ((buf[offset] & 0x80) === 0) {
        // Positive number
        return module.exports.readUInt64BE(buf, offset);
    }

    // Negative number
    const lowerValue = -buf.readInt32BE(offset + 4);
    let bnUpperValue;

    if (lowerValue === 0) {
        bnUpperValue = new BigNumber(-buf.readInt32BE(offset));
    }
    else {
        bnUpperValue = new BigNumber('ffffffff', 16).minus(buf.readUInt32BE(offset));
    }

    return bnUpperValue.times(0x100000000).plus(lowerValue).negated().toNumber();
};


src/MessageReceipt.js
const bitcoinLib = require('bitcoinjs-lib');
const varint = require('varint');
const MessageEnvelope = require('./MessageEnvelope');
const Util = require('./Util');

const structID = 0x52;
const initVersion = 0x01;
const currentVersion = initVersion;
const headerFixedLength = 2 + (20 * 2) + 8;
const minHeaderLength = 1 + headerFixedLength + 3;
const minSignDataLength = 1 + 1 + 33;

class MessageReceipt {
    /**
     * Class constructor
     * @param {Object} rcptInfo Object containing information describing the message receipt to create, with the
     *                           following properties:
     *                           - msgInfo (MessageEnvelope|Object) Info about the message for which this receipt is issued.
     *                               Can be either an instance of MessageEnvelope or a literal object with the following properties:
     *                               - senderPubKeyHash (String|Buffer) The public key hash of the Catenis device that sent the message
     *                               - receiverPubKeyHash (String|Buffer) The public key hash of the Catenis device to which the message was sent
     *                           - timestamp (Integer) (optional) The timestamp (milliseconds from Unix's epoch) when the message receipt was created.
     *                               If not specified, the current time is used
     *                           - msgEnvCid (String|Buffer|CID) The IPFS CID of the message envelope of the message for which this receipt is issued
     */
    constructor(rcptInfo) {
        // Validate parameter
        if (typeof rcptInfo !== 'object' || rcptInfo === null) {
            throw new Error('Missing or invalid `rcptInfo` parameter');
        }

        const rcptInfoErrors = [];
        let msgInfoError = false;

        if (rcptInfo.msgInfo instanceof MessageEnvelope) {
            this.msgEnv = rcptInfo.msgInfo;

            if (!isValidMsgEnvelope(this.msgEnv)) {
                rcptInfoErrors.push('inconsistent `msgInfo` property: message does not require receipt');
            }
            else {
                this.senderPubKeyHash = this.msgEnv.senderPubKeyHash;
                this.receiverPubKeyHash = this.msgEnv.receiverPubKeyHash;
            }
        }
        else if (typeof rcptInfo.msgInfo !== 'object' || rcptInfo.msgInfo === null) {
            rcptInfoErrors.push('missing or invalid `msgInfo` property');
            msgInfoError = true;
        }

        if (!this.msgEnv && !msgInfoError) {
            this.senderPubKeyHash = Util.validatePubKeyHash(rcptInfo.msgInfo.senderPubKeyHash);

            if (!this.senderPubKeyHash) {
                rcptInfoErrors.push('missing or invalid `msgInfo.senderPubKeyHash` property');
            }

            this.receiverPubKeyHash = Util.validatePubKeyHash(rcptInfo.msgInfo.receiverPubKeyHash);

            if (!this.receiverPubKeyHash) {
                rcptInfoErrors.push('missing or invalid `msgInfo.receiverPubKeyHash` property');
            }
        }

        if (rcptInfo.timestamp) {
            if (!Util.isValidTimestamp(rcptInfo.timestamp)) {
                rcptInfoErrors.push('invalid `timestamp` property');
            }
            else {
                this.timestamp = rcptInfo.timestamp;
            }
        }
        else {
            this.timestamp = Date.now();
        }

        this.msgEnvCid = Util.validateCid(rcptInfo.msgEnvCid);

        if (!this.msgEnvCid) {
            rcptInfoErrors.push('missing or invalid `msgEnvCid` property');
        }

        if (rcptInfoErrors.length > 0) {
            throw new Error('Invalid `rcptInfo` parameter: ' + rcptInfoErrors.join('; '));
        }

        // Encode data structure
        const fixedHeader = Buffer.alloc(headerFixedLength);
        let offset = 0;

        offset = fixedHeader.writeUInt8(structID, offset);
        this.version = currentVersion;
        offset = fixedHeader.writeUInt8(this.version, offset);
        offset += this.senderPubKeyHash.copy(fixedHeader, offset);
        offset += this.receiverPubKeyHash.copy(fixedHeader, offset);
        Util.writeInt64BE(this.timestamp, fixedHeader, offset);

        const header = Buffer.concat([fixedHeader, this.msgEnvCid.bytes]);

        // Add header length prefix
        let lenPrefixLength = 1;

        while (varint.encodingLength(header.byteLength + lenPrefixLength) !== lenPrefixLength) {
            lenPrefixLength++;
        }

        this.buffer = this.header = Buffer.concat([Buffer.from(varint.encode(header.byteLength + lenPrefixLength)), header]);

        this.hasSignature = false;
    }

    get hex() {
        return this.buffer.toString('hex');
    }

    get base64() {
        return this.buffer.toString('base64');
    }

    get isMessageChecked() {
        return !!this.msgEnv;
    }

    get isSigned() {
        return this.hasSignature;
    }

    checkMessage(msgEnv) {
        // Reset error
        this.checkMessageError = undefined;

        if (this.msgEnv) {
            this.checkMessageError = 'Message already checked for receipt';
        }

        if (!this.checkMessageError && !(msgEnv instanceof MessageEnvelope)) {
            this.checkMessageError = 'Invalid message: not an instance of MessageEnvelope';
        }

        if (!this.checkMessageError && !isValidMsgEnvelope(msgEnv)) {
            this.checkMessageError = 'Invalid message: it does not require receipt';
        }

        if (!this.checkMessageError && (!msgEnv.senderPubKeyHash.equals(this.senderPubKeyHash) || !msgEnv.receiverPubKeyHash.equals(this.receiverPubKeyHash))) {
            this.checkMessageError = 'Invalid message: it does not match sender and/or receiver';
        }

        if (!this.checkMessageError) {
            // Message successfully checked
            this.msgEnv = msgEnv;
            return true;
        }
        else {
            return false;
        }
    }

    sign(keyPair) {
        if (!this.hasSignature) {
            if (!keyPair.compressed) {
                throw new Error('Invalid public key format; it should be compressed');
            }

            // Make sure that passed in key pair matches message receiver's public key hash
            if (!bitcoinLib.crypto.hash160(keyPair.publicKey).equals(this.receiverPubKeyHash)) {
                throw new Error('Passed in key pair does not match message receiver\'s public key hash');
            }

            this.signature = keyPair.sign(bitcoinLib.crypto.sha256(this.header));
            this.receiverPubKey = keyPair.publicKey;

            this.buffer = Buffer.concat([
                this.header,
                Buffer.from(varint.encode(this.signature.byteLength)),
                this.signature,
                this.receiverPubKey
            ]);

            this.hasSignature = true;
        }
    }

    verifySignature() {
        if (this.hasSignature) {
            try {
                const keyPair = Util.keyPairFromPublicKey(this.receiverPubKey);

                return keyPair.verify(bitcoinLib.crypto.sha256(this.header), this.signature);
            }
            catch (err) {
                return false;
            }
        }
    }

    static fromBuffer(buf) {
        if (!Buffer.isBuffer(buf)) {
            throw new TypeError('Invalid argument type; expected Buffer');
        }

        // Validate data structure header first
        if (buf.byteLength < minHeaderLength) {
            throw new Error('Data buffer too short');
        }

        let offset = 0;
        let headerLength;

        try {
            headerLength = varint.decode(buf, offset);
        }
        catch (err) {
            throw new Error('Invalid header length');
        }

        const lenPrefixLength = varint.decode.bytes;

        if (headerLength - lenPrefixLength < minHeaderLength - 1 || headerLength > buf.byteLength) {
            throw new Error('Inconsistent header length');
        }

        offset += lenPrefixLength;
        const structIdByte = buf.readUInt8(offset++);
        const versionByte = buf.readUInt8(offset++);
        const senderPubKeyHash = buf.slice(offset, offset + 20);
        offset += 20;
        const receiverPubKeyHash = buf.slice(offset, offset + 20);
        offset += 20;
        const timestamp = Util.readInt64BE(buf, offset);
        offset += 8;
        const msgEnvCid = buf.slice(offset, headerLength);
        offset = headerLength;

        const headerErrors = [];

        if (!isValidStructIdByte(structIdByte)) {
            headerErrors.push('invalid structure ID');
        }

        if (!isValidVersionByte(versionByte)) {
            headerErrors.push('invalid version byte');
        }

        if (!Util.isValidTimestamp(timestamp)) {
            headerErrors.push('invalid timestamp');
        }

        const rcptInfo = {
            msgEnvCid: Util.validateCid(msgEnvCid)
        };

        if (!rcptInfo.msgEnvCid) {
            headerErrors.push('invalid message content CID');
        }

        if (headerErrors.length > 0) {
            throw new Error('Inconsistent header data: ' + headerErrors.join('; '));
        }

        rcptInfo.msgInfo = {
            senderPubKeyHash: senderPubKeyHash,
            receiverPubKeyHash: receiverPubKeyHash
        };
        rcptInfo.timestamp = timestamp;

        const msgRcpt = new MessageReceipt(rcptInfo);

        // Check if signature is present in data structure
        const bufSig = buf.slice(offset);

        if (bufSig.byteLength > 0) {
            // Validate signature part of data structure
            if (bufSig.byteLength < minSignDataLength) {
                throw new Error('Signature data too short');
            }

            offset = 0;
            let signDataLength;

            try {
                signDataLength = varint.decode(bufSig, offset);
            }
            catch (err) {
                throw new Error('Invalid signature length');
            }

            const lenPrefixLength = varint.decode.bytes;

            if (signDataLength < 1) {
                throw new Error('Inconsistent signature length');
            }

            msgRcpt.signature = bufSig.slice(lenPrefixLength, lenPrefixLength + signDataLength);

            if (msgRcpt.signature.byteLength < signDataLength) {
                throw new Error('Inconsistent signature data: signature shorter than expected');
            }

            offset += lenPrefixLength + signDataLength;
            msgRcpt.receiverPubKey = bufSig.slice(offset);

            if (msgRcpt.receiverPubKey.byteLength === 0) {
                throw new Error('Inconsistent signature data: missing public key');
            }

            // Validate message receiver public key
            let keyPair;

            try {
                keyPair = Util.keyPairFromPublicKey(msgRcpt.receiverPubKey);
            }
            catch (err) {
                throw new Error('Inconsistent signature data: invalid public key')
            }

            if (!keyPair.compressed) {
                throw new Error('Inconsistent signature data: invalid public key format; it should be compressed');
            }

            if (!bitcoinLib.crypto.hash160(keyPair.publicKey).equals(msgRcpt.receiverPubKeyHash)) {
                throw new Error('Inconsistent signature data: public key does not match message receiver\'s public key hash');
            }

            // Add signature data
            msgRcpt.buffer = Buffer.concat([
                msgRcpt.header,
                Buffer.from(varint.encode(msgRcpt.signature.byteLength)),
                msgRcpt.signature,
                msgRcpt.receiverPubKey
            ]);

            msgRcpt.hasSignature = true;
        }

        return msgRcpt;
    }

    static fromHex(hex) {
        return MessageReceipt.fromBuffer(Buffer.from(hex, 'hex'));
    }

    static fromBase64(base64) {
        return MessageReceipt.fromBuffer(Buffer.from(base64, 'base64'));
    }
}

function isValidMsgEnvelope(msgEnv) {
    return msgEnv.msgType === MessageEnvelope.msgType.sendMessage && msgEnv.isMessageWithReadConfirmation;
}

function isValidStructIdByte(byte) {
    return byte === structID;
}

function isValidVersionByte(byte) {
    return byte >= initVersion && byte <= currentVersion;
}

module.exports = MessageReceipt;

src/BatchDocument.js
const util = require('util');
const zlib = require('zlib');
const bitcoinLib = require('bitcoinjs-lib');
const merkle = require('merkle-lib');
const merkleProof = require('merkle-lib/proof');
const MessageEnvelope = require('./MessageEnvelope');
const MessageReceipt = require('./MessageReceipt');
const Util = require('./Util');

const initialVersion = 0;
const latestVersion = 1;
const hashFunctions = [
    // Initial version (ver. 0)
    {
        leaf(node) {
            return Buffer.concat([Buffer.from([0x00]), node]);
        },
        internalNode(node) {
            return Buffer.concat([Buffer.from([0x01]), bitcoinLib.crypto.hash256(node)]);
        }
    },
    // Version 1
    {
        leaf(node) {
            return bitcoinLib.crypto.sha256(Buffer.concat([Buffer.from([0x00]), node]));
        },
        internalNode(node) {
            return bitcoinLib.crypto.sha256(Buffer.concat([Buffer.from([0x01]), node]));
        }
    }
];

class BatchDocument {
    /**
     * Class constructor
     * @param {[Object]} entries An array of objects containing information describing the message data (either message envelopes or message receipts)
     *                            to be included in the batch, with the following properties:
     *                            - msgInfo (MessageEnvelope|MessageReceipt|Object) Information about the message the message envelope or receipt of which is to be added to the batch.
     *                                Can be either an instance of MessageEnvelope, and instance of MessageReceipt or a literal object with the following properties:
     *                                - senderPubKeyHash (String|Buffer) The public key hash of the Catenis device that sent the message
     *                                - receiverPubKeyHash (String|Buffer) (optional) The public key hash of the Catenis device to which the message was sent
     *                            - msgDataCid (String|Buffer|CID) The IPFS CID of the message envelope or receipt to add to the batch
     * @param {Number} [version] Version of batch document data structure to use
     */
    constructor(entries, version = latestVersion) {
        if (!Array.isArray(entries)) {
            entries = [entries];
        }

        // Validate parameter
        if (entries.length === 0 || (entries.length === 1 && entries[0] === undefined)) {
            throw new Error('Missing or invalid `entries` parameter');
        }

        if (!isValidVersion(version)) {
            throw new Error('Invalid `version` parameter');
        }

        this.entries = [];
        this.version = version;
        this.setMsgDataCids = new Set();
        this.mapSenderPubKeyHashes = new Map();
        this.mapReceiverPubKeyHashes = new Map();

        for (let idx = 0, numEntries = entries.length; idx < numEntries; idx++) {
            const result = validateEntry(entries[idx]);
            let error;
            let strCid;

            if (!result.success) {
                error = result.error;
            }
            else if (this.setMsgDataCids.has(strCid = result.entry.msgDataCid.toString())) {
                error = 'duplicate `msgDataCid` property value';
            }

            if (error) {
                throw new Error(util.format('Invalid entry #%d: %s', idx + 1, error));
            }

            this.entries.push(result.entry);
            this.setMsgDataCids.add(strCid);
            addMapListItem(this.mapSenderPubKeyHashes, result.entry.senderPubKeyHash.toString('base64'), idx);

            if (result.entry.receiverPubKeyHash) {
                addMapListItem(this.mapReceiverPubKeyHashes, result.entry.receiverPubKeyHash.toString('base64'), idx);
            }
        }

        this.msgDataCids = Array.from(this.setMsgDataCids);

        // Instantiate Merkle tree
        this.hashFunction = hashFunctions[this.version];
        this.tree = merkle(this.entries.map(entry => this.hashFunction.leaf(Buffer.from(entry.msgDataCid.bytes.buffer, entry.msgDataCid.bytes.byteOffset, entry.msgDataCid.bytes.byteLength))), this.hashFunction.internalNode);

        // Assemble batch document
        this.doc = {
            version: this.version,
            msgData: this.msgDataCids,
            senders: Array.from(this.mapSenderPubKeyHashes),
            receivers: Array.from(this.mapReceiverPubKeyHashes),
            merkleRoot: this.merkleRoot.toString('base64')
        };

        if (version === initialVersion) {
            delete this.doc.version;
        }
    }

    get merkleRoot() {
        return this.tree[this.tree.length - 1];
    }

    get hex() {
        if (this.buffer) {
            return this.buffer.toString('hex');
        }
    }

    get base64() {
        if (this.buffer) {
            return this.buffer.toString('base64');
        }
    }

    get indicesEntryToCheckMessageData() {
        return this.entries.reduce((indices, entry, idx) => {
            if (!entry.msgData) {
                indices.push(idx)
            }

            return indices;
        }, []);
    }

    get isBuilt() {
        return !!this.buffer;
    }

    build() {
        if (!this.buffer) {
            const json = JSON.stringify(this.doc);

            this.buffer = zlib.gzipSync(Buffer.from(json));
        }
    }

    isAllMessageDataChecked() {
        return this.entries.every(entry => entry.msgData);
    }

    // Note: each listMsgData item (MessageEnvelope or MessageReceipt) should correspond to a batch document entry.
    //      Entries for which the message data has already been checked should have the corresponding item set to undefined
    checkMessageData(listMsgData) {
        if (!Array.isArray(listMsgData)) {
            listMsgData = [listMsgData];
        }

        const numEntries = this.entries.length;

        if (listMsgData.length !== numEntries) {
            throw new Error('Number of message data items do not match number of entries');
        }

        // Reset error
        this.listCheckMsgDataError = [];
        let hasError = false;

        for (let idx = 0; idx < numEntries; idx++) {
            const entry = this.entries[idx];
            const msgData = listMsgData[idx];
            let error;

            if (entry.msgData) {
                if (msgData !== undefined) {
                    error = 'Message data already checked for entry';
                }
            }
            else {
                error = checkMsgDataItem(msgData, entry);
            }

            if (error) {
                this.listCheckMsgDataError.push(error);
                hasError = true;
            }
            else {
                entry.msgData = msgData;
                this.listCheckMsgDataError.push(undefined);
            }
        }

        return !hasError;
    }

    isMessageDataInBatch(msgDataCid) {
        msgDataCid = Util.validateCid(msgDataCid);

        if (!msgDataCid) {
            throw new Error('Invalid message data (envelope or receipt) CID');
        }

        const proof = merkleProof(this.tree, this.hashFunction.leaf(Buffer.from(msgDataCid.bytes.buffer, msgDataCid.bytes.byteOffset, msgDataCid.bytes.byteLength)));

        if (proof) {
            return merkleProof.verify(proof, this.hashFunction.internalNode);
        }

        return false;
    }

    static fromBuffer(buf) {
        if (!Buffer.isBuffer(buf)) {
            throw new TypeError('Invalid argument type; expected Buffer');
        }

        // Try to decompress buffer
        let uncompressedBuf;

        try {
            uncompressedBuf = zlib.gunzipSync(buf);
        }
        catch (err) {
            throw new Error('Data is not compressed as expected');
        }

        // Try to parse JSON
        let json;

        try {
            json = JSON.parse(uncompressedBuf.toString());
        }
        catch (err) {}

        if (json === undefined || typeof json !== 'object' || json === null) {
            throw new Error('Data is not a valid JSON object');
        }

        // Check object structure
        if (!checkObjectProperties(json, {
            'version?': 'number',
            msgData: 'array',
            senders: 'array',
            receivers: 'array',
            merkleRoot: 'string'
        })) {
            throw new Error('Invalid batch document object');
        }

        let version;

        if ('version' in json) {
            if (!isValidVersion(json.version) || json.version === initialVersion) {
                throw new Error('Invalid `version` property of batch document object');
            }

            version = json.version;
        }
        else {
            // No version property present. Set version as initial version
            version = initialVersion;
        }

        if (json.msgData.length === 0) {
            throw new Error('Invalid `msgData` property of batch document object');
        }

        let sendersPubKeyHashes;
        
        try {
            sendersPubKeyHashes = new Map(json.senders);
        }
        catch (err) {
            throw new Error('Invalid `senders` property of batch document object');
        }

        let receiversPubKeyHashes;

        try {
            receiversPubKeyHashes = new Map(json.receivers);
        }
        catch (err) {
            throw new Error('Invalid `receivers` property of batch document object');
        }
        
        // Assemble batch document entries
        const entries = json.msgData.map(msgData => {
            return {
                msgInfo: {},
                msgDataCid: msgData
            }
        });
        const numEntries = entries.length;

        // Add sender public key hashes
        for (let [hash, indices] of sendersPubKeyHashes) {
            if (!Array.isArray(indices)) {
                throw new Error('Invalid list of indices for sender');
            }

            indices.forEach(idx => {
                if (!Number.isInteger(idx) || idx < 0 || idx >= numEntries) {
                    throw new Error('Invalid index for sender');
                }

                const entry = entries[idx];

                if ('senderPubKeyHash' in entry.msgInfo) {
                    throw new Error('Duplicate sender for entry');
                }

                entry.msgInfo.senderPubKeyHash = hash;
            })
        }

        // Add receiver public key hashes
        for (let [hash, indices] of receiversPubKeyHashes) {
            if (!Array.isArray(indices)) {
                throw new Error('Invalid list of indices for receiver');
            }

            indices.forEach(idx => {
                if (!Number.isInteger(idx) || idx < 0 || idx >= numEntries) {
                    throw new Error('Invalid index for receiver');
                }

                const entry = entries[idx];

                if ('receiverPubKeyHash' in entry.msgInfo) {
                    throw new Error('Duplicate receiver for entry')
                }

                entry.msgInfo.receiverPubKeyHash = hash;
            })
        }

        let batchDoc;

        try {
            batchDoc = new BatchDocument(entries, version);
        }
        catch (err) {
            throw new Error('Invalid batch document entries');
        }

        if (batchDoc.merkleRoot.toString('base64') !== json.merkleRoot) {
            throw new Error('Inconsistent Merkle root hash in batch document');
        }

        // Everything is OK. Finalize batch document and return it
        batchDoc.buffer = buf;

        return batchDoc;
    }

    static fromHex(hex) {
        return BatchDocument.fromBuffer(Buffer.from(hex, 'hex'));
    }

    static fromBase64(base64) {
        return BatchDocument.fromBuffer(Buffer.from(base64, 'base64'));
    }
}

function isValidVersion(n) {
    return Number.isInteger(n) && n >= initialVersion && n <= latestVersion;
}

function validateEntry(entry) {
    const result = {};
    const validatedEntry = {};
    const errors = [];

    if (typeof entry !== 'object' || entry === null) {
        errors.push('invalid entry type');
    }
    else {
        let msgInfoError = false;

        if (entry.msgInfo instanceof MessageEnvelope || entry.msgInfo instanceof MessageReceipt) {
            validatedEntry.msgData = entry.msgInfo;

            validatedEntry.senderPubKeyHash = validatedEntry.msgData.senderPubKeyHash;
            validatedEntry.receiverPubKeyHash = validatedEntry.msgData.receiverPubKeyHash;
        }
        else if (typeof entry.msgInfo !== 'object' || entry.msgInfo === null) {
            errors.push('missing or invalid `msgInfo` property');
            msgInfoError = true;
        }

        if (!validatedEntry.msgData && !msgInfoError) {
            validatedEntry.senderPubKeyHash = Util.validatePubKeyHash(entry.msgInfo.senderPubKeyHash);

            if (!validatedEntry.senderPubKeyHash) {
                errors.push('missing or invalid `msgInfo.senderPubKeyHash` property');
            }

            if ('receiverPubKeyHash' in entry.msgInfo) {
                validatedEntry.receiverPubKeyHash = Util.validatePubKeyHash(entry.msgInfo.receiverPubKeyHash);

                if (!validatedEntry.receiverPubKeyHash) {
                    errors.push('invalid `msgInfo.receiverPubKeyHash` property');
                }
            }
        }

        validatedEntry.msgDataCid = Util.validateCid(entry.msgDataCid);

        if (!validatedEntry.msgDataCid) {
            errors.push('missing or invalid `msgDataCid` property');
        }
    }

    if (errors.length > 0) {
        result.success = false;
        result.error = errors.join('; ');
    }
    else {
        result.success = true;
        result.entry = validatedEntry;
    }
    
    return result;
}

function addMapListItem(map, key, item) {
    if (!map.has(key)) {
        map.set(key, [item]);
    }
    else {
        map.get(key).push(item);
    }
}

function checkMsgDataItem(msgData, entry) {
    let error;

    if (!(msgData instanceof MessageEnvelope || msgData instanceof MessageReceipt)) {
        error = 'Invalid message data: not an instance of MessageEnvelope nor MessageReceipt';
    }

    if (!error && (!msgData.senderPubKeyHash.equals(entry.senderPubKeyHash) || (entry.receiverPubKeyHash && !msgData.receiverPubKeyHash || !entry.receiverPubKeyHash && msgData.receiverPubKeyHash || msgData.receiverPubKeyHash && !msgData.receiverPubKeyHash.equals(entry.receiverPubKeyHash)))) {
        error = 'Invalid message data: it does not match sender and/or receiver';
    }

    return error;
}

// Argument:
//  props (Object) A map where key is property name and value the expected property type
function checkObjectProperties(obj, props) {
    let requiredPropsFound = 0;

    let error = Object.keys(obj).some(key => {
        let prop = key;

        if ((prop in props) || ((prop += '?') in props)) {
            if (!((props[prop] === 'array' && Array.isArray(obj[key])) || props[prop] === typeof obj[key])) {
                // Key has an unexpected type.
                //  Stop iteration indicating failure
                return true;
            }

            // Check if it is not an optional property
            if (!prop.endsWith('?')) {
                requiredPropsFound++;
            }

            return false;
        }
        else {
            // Key is not one of the expected props.
            //  Stop iteration indicating failure
            return true;
        }
    });

    if (!error) {
        const totalRequiredProps = Object.keys(props).reduce((count, prop) => {
            if (!prop.endsWith('?')) {
                count++;
            }

            return count;
        }, 0);

        error = requiredPropsFound !== totalRequiredProps;
    }

    return !error;
}

module.exports = BatchDocument;